src/Facet/Context.hs

```diff
@@ -12,55 +11,47 @@ module Facet.Context
 ) where
 
 import qualified Control.Effect.Empty as E
+import           Data.Foldable (find, toList)
+import           Facet.Core.Pattern
 import           Facet.Core.Type
+import qualified Facet.Env as Env
 import           Facet.Name
 import qualified Facet.Snoc as S
+import           Facet.Syntax
 import           Facet.Usage
 import           GHC.Stack
-import           Prelude hiding (lookup, zipWith)
+import           Prelude hiding (lookup)
 
-newtype Context = Context { elems :: S.Snoc Binding }
-
-data Binding = Binding
-  { name     :: Name
-  , quantity :: Quantity
-  , type'    :: Either Kind Type
-  }
+newtype Context = Context { elems :: S.Snoc (Quantity, Pattern (Name ::: Classifier)) }
```

Comment:
The richer structure of `Context`s allows us to express that the quantity for an argument associates to the entire pattern, not individual variables.